File: org.kevoree.modeling.microframework/src/main/java/org/kevoree/modeling/api/data/MemoryKDataBase.java
package org.kevoree.modeling.api.data;

import org.kevoree.modeling.api.Callback;
import org.kevoree.modeling.api.ThrowableCallback;

import java.util.HashMap;

public class MemoryKDataBase implements KDataBase {

    private final HashMap<String, String> backend = new HashMap<String, String>();

    public static boolean DEBUG = false;

    @Override
    public void put(String[][] payloads, Callback<Throwable> callback) {
        for (int i = 0; i < payloads.length; i++) {
            backend.put(payloads[i][0], payloads[i][1]);
            if (DEBUG) {
                System.out.println("PUT " + payloads[i][0] + "->" + payloads[i][1]);
            }
        }
        callback.on(null);
    }

    @Override
    public void get(String[] keys, ThrowableCallback<String[]> callback) {
        String[] values = new String[keys.length];
        for (int i = 0; i < keys.length; i++) {
            values[i] = backend.get(keys[i]);
            if (DEBUG) {
                System.out.println("GET " + keys[i] + "->" + values[i]);
            }
        }
        callback.on(values, null);
    }

    @Override
    public void remove(String[] keys, Callback<Throwable> callback) {
        for (int i = 0; i < keys.length; i++) {
            backend.remove(keys[i]);
        }
        callback.on(null);
    }

    @Override
    public void commit(Callback<Throwable> callback) {
        callback.on(null);//noop
    }

    @Override
    public void connect(Callback<Throwable> callback) {
        if (callback != null) {
            callback.on(null);
        }
    }

    @Override
    public void close(Callback<Throwable> callback) {
        backend.clear();
    }

}

File: org.kevoree.modeling.microframework.typescript/src/main/ts/org/kevoree/modeling/api/data/MemoryKDataBase.ts
///<reference path="../Callback.ts"/>
///<reference path="../ThrowableCallback.ts"/>
///<reference path="../../../../../java/util/JUHashMap.ts"/>

class MemoryKDataBase implements KDataBase {

  private backend: JUHashMap<string, string> = new JUHashMap<string, string>();

  public put(payloads: string[][], callback: Callback<Throwable>): void {
    for (var i: number = 0; i < payloads.length; i++) {
      this.backend.put(payloads[i][0], payloads[i][1]);
    }
    callback.on(null);
  }

  public get(keys: string[], callback: ThrowableCallback<string[]>): void {
    var values: string[] = new Array();
    for (var i: number = 0; i < keys.length; i++) {
      values[i] = this.backend.get(keys[i]);
    }
    callback.on(values, null);
  }

  public remove(keys: string[], callback: Callback<Throwable>): void {
    for (var i: number = 0; i < keys.length; i++) {
      this.backend.remove(keys[i]);
    }
    callback.on(null);
  }

  public commit(callback: Callback<Throwable>): void {
  }

  public close(callback: Callback<Throwable>): void {
    this.backend.clear();
  }

}

